crates/red_knot_python_semantic/resources/mdtest/statically-known-branches.md

```diff
@@ -0,0 +1,1037 @@
+
+```py
+def iterable() -> list[object]: ...
+
+x = 1
+
+if True:
+    for _ in iterable():
+        x = 2
+    else:
+        x = 3
+else:
+    x = 4
+
+reveal_type(x)  # revealed: Literal[3]
+```
+
+##### `for` loop with `break` inside `if True`
+
+```py
+def iterable() -> list[object]: ...
+
+x = 1
+
+if True:
+    x = 2
+    for _ in iterable():
+        x = 3
+        break
+    else:
+        x = 4
+else:
+    x = 5
+
+reveal_type(x)  # revealed: Literal[3, 4]
+```
+
+## If expressions
+
+See also: tests in [expression/if.md](expression/if.md).
+
+### Always true
+
+```py
+x = 1 if True else 2
+
+reveal_type(x)  # revealed: Literal[1]
+```
+
+### Always false
+
+```py
+x = 1 if False else 2
+
+reveal_type(x)  # revealed: Literal[2]
+```
+
+## Boolean expressions
+
+### Always true, `or`
+
+```py
+(x := 1) or (x := 2)
+
+reveal_type(x)  # revealed: Literal[1]
+```
+
+### Always true, `and`
+
+```py
+(x := 1) and (x := 2)
+
+reveal_type(x)  # revealed: Literal[2]
+```
+
+### Always false, `or`
+
+```py
+(x := 0) or (x := 2)
+
+reveal_type(x)  # revealed: Literal[2]
+```
+
+### Always false, `and`
+
+```py
+(x := 0) and (x := 2)
+
+reveal_type(x)  # revealed: Literal[0]
+```
+
+## While loops
+
+### Always false
+
+```py
+x = 1
+
+while False:
+    x = 2
+
+reveal_type(x)  # revealed: Literal[1]
+```
+
+### Always true
+
+```py
+x = 1
+
+while True:
+    x = 2
+    break
+
+reveal_type(x)  # revealed: Literal[2]
+```
+
+### Ambiguous
+
+Make sure that we still infer the combined type if the condition is not statically known:
+
+```py
+def flag() -> bool: ...
+
+x = 1
+
+while flag():
+    x = 2
+
+reveal_type(x)  # revealed: Literal[1, 2]
+```
+
+### `while` ... `else`
+
+#### `while False`
+
+```py
+while False:
+    x = 1
+else:
+    x = 2
+
+reveal_type(x)  # revealed: Literal[2]
+```
+
+#### `while True`
+
+```py
+while True:
+    x = 1
+    break
+else:
+    x = 2
+
+reveal_type(x)  # revealed: Literal[1]
+```
+
+## `match` statements
+
+### Single-valued types, always true
+
+```py
+x = 1
+
+match "a":
+    case "a":
+        x = 2
+    case "b":
+        x = 3
+
+reveal_type(x)  # revealed: Literal[2]
+```
+
+### Single-valued types, always false
+
+```py
+x = 1
+
+match "something else":
+    case "a":
+        x = 1
+    case "b":
+        x = 2
+
+reveal_type(x)  # revealed: Literal[1]
+```
+
+### Single-valued types, with wildcard pattern
+
+This is a case that we can not handle at the moment. Our reasoning about match patterns is too
+local. We can infer that the `x = 2` binding is unconditionally visible. But when we traverse all
+bindings backwards, we first see the `x = 3` binding which is also visible. At the moment, we do not
+mark it as *unconditionally* visible to avoid blocking off previous bindings (we would infer
+`Literal[3]` otherwise).
+
+```py
+x = 1
+
+match "a":
+    case "a":
+        x = 2
+    case _:
+        x = 3
+
+# TODO: ideally, this should be Literal[2]
+reveal_type(x)  # revealed: Literal[2, 3]
+```
```

Comment:
I just realized that this is problematic in cases with multiple "always True" patterns. At runtime, we'll take the first of these. But in our current implementation here, since we always walk definitions backwards, we'll take the last instead:
```rs
x = 1

match "a":
    case "a" if 1 == 1:
        x = 2
    case "a":
        x = 3

reveal_type(x)
```

Before I attempt to fix this, it would be good to get some feedback on whether or not we want static-truthiness checking in `match` statements at all. I thought it was a nice stretch-task, but maybe it causes more trouble than it's worth. The examples presented here are not particularly useful, but there might be some use cases involving enums? Not sure.

Reply:
> I'm curious why this seems to affect match statements with multiple always-matching cases, but not `elif` chains with multiple always-true branches?

It currently works for `elif` chains but not for match patterns because of negative constraints. The way it currently works is that we record all constraints (including negative ones) when a binding is created:

```py
if flag:
    x = 1  # conditional on flag
elif True:
    x = 2  # conditional on ~flag and True
elif True:
    x = 3  # conditional on ~flag, ~True and True
```

The `x = 3` binding is therefore *not* unconditionally visible, which currently leads us to correctly infer `Literal[1, 2]` for this case.

But now that I write it out like this, I also realize that this reveals a problem with my approach. Consider this example:

```py
x = 1

if flag:
    x = 2  # conditional on flag
elif True:
    x = 3  # conditional on ~flag, True
```

Here, we currently infer `Literal[1, 2, 3]`, while it would be more correct to infer `Literal[2, 3]`. The problem is that we have the negative `~flag` condition in the `x = 3` binding, which prevents us from detecting it as "unconditionally visible". And if we *would* detect it as unconditionally visible, we would run into the problem you mentioned. We would block off any previous bindings and incorrectly infer `Literal[3]`.

We could clearly do better.

I'm currently thinking that there are two problems:
* It seems to me like the set of constraints that we apply for static truthiness analysis needs to be different from the set of constraints that we apply for narrowing. For narrowing, we need that `~flag` constraint. But for static truthiness analysis, we don't care if `flag` has ambiguous truthiness or not, we always take an `elif True` branch.
* When iterating the list of bindings in reverse, we should not stop at the first "unconditionally visible" binding, we need to go back to the earliest possible "unconditionally visible" binding.